hal/ia32/timer.c

```diff
@@ -16,80 +16,418 @@
 
 #include "hal/timer.h"
 #include "hal/cpu.h"
+#include "init.h"
 #include "hal/interrupts.h"
 #include "hal/spinlock.h"
 #include "ia32.h"
 
+#define PIT_FREQUENCY 1193 /* kHz */
+
+#define PIT_BCD                0
+#define PIT_CHANNEL_0          (0 << 6)
+#define PIT_CHANNEL_1          (1 << 6)
+#define PIT_CHANNEL_2          (2 << 6)
+#define PIT_ACCESS_BOTH        (3 << 4)
+#define PIT_OPERATING_ONE_SHOT (0 << 1)
+#define PIT_OPERATING_RATE_GEN (2 << 1)
+
+#define LAPIC_TIMER_ONE_SHOT        0
+#define LAPIC_TIMER_DEFAULT_DIVIDER 3 /* 3 means 8 (1 << 3) */
+
+/* 64 bit registers, access must be aligned
+   Trying to use exclusive-access mechanisms (ex. lock mov, xchg, etc.) is undefined */
+#define HPET_ID         0x00
+#define HPET_CONFIG     0x10
+#define HPET_IRQ_STATUS 0x20
+#define HPET_COUNTER    0xf0
+
+#define HPET_ID_LEGACY_CAPABLE           (1u << 15)
+#define HPET_LEGACY_TMR1_IRQ             8
+#define HPET_CONFIG_TMR_IRQ_EN           (1u << 2)
+#define HPET_CONFIG_TMR_PERIODIC         (1u << 3)
+#define HPET_CONFIG_TMR_CAN_BE_PERIODIC  (1u << 4)
+#define HPET_CONFIG_TMR_PERIODIC_CAN_SET (1u << 6)
+#define HPET_CONFIG_TMR_32BIT_MODE       (1u << 8)
+
+typedef enum { timer_unknown,
+	timer_pit,
+	timer_lapic,
+	timer_hpet } timerType_t;
```

Comment:
clang-format is mistaken
```suggestion
typedef enum { timer_unknown, timer_pit, timer_lapic, timer_hpet } timerType_t;
```